pkg/controller/argocd/util.go

```diff
@@ -1064,10 +1067,98 @@ func containsString(arr []string, s string) bool {
 			if _, ok := e.MetaNew.GetLabels()[common.ArgoCDManagedByLabel]; ok {
 				return true
 			}
+			// This checks if the old meta had the label, if it did, delete the RBACs for the namespace
+			// which were created when the label was added to the namespace.
+			if ns, ok := e.MetaOld.GetLabels()[common.ArgoCDManagedByLabel]; ok && ns != "" {
+				if err := deleteRBACsForNamespace(ns, e.MetaOld.GetName()); err != nil {
+					log.Error(err, fmt.Sprintf("failed to delete RBACs for namespace: %s", e.MetaOld.GetName()))
+				} else {
+					log.Info(fmt.Sprintf("Successfully removed the RBACs for namespace: %s", e.MetaOld.GetName()))
+				}
+			}
 			return false
 		},
 	}
 }
+
+// deleteRBACsForNamespace deletes the RBACs when the label from the namespace is removed.
+func deleteRBACsForNamespace(ownerNS, sourceNS string) error {
+	log.Info(fmt.Sprintf("Removing the RBACs created for the namespace: %s", sourceNS))
+	cfg, err := config.GetConfig()
+	if err != nil {
+		log.Error(err, "unable to get k8s config")
+		return err
+	}
+
+	k8sClient, err := kubernetes.NewForConfig(cfg)
+	if err != nil {
+		log.Error(err, "unable to create k8s client")
+		return err
+	}
+
+	// List all the roles created for ArgoCD using the label selector
+	labelSelector := metav1.LabelSelector{MatchLabels: map[string]string{common.ArgoCDKeyPartOf: common.ArgoCDAppName}}
+	roles, err := k8sClient.RbacV1().Roles(sourceNS).List(context.TODO(), metav1.ListOptions{LabelSelector: labels.Set(labelSelector.MatchLabels).String()})
+	if err != nil {
+		log.Error(err, fmt.Sprintf("failed to list roles for namespace: %s", sourceNS))
+		return err
+	}
+
+	// Delete all the retrieved roles
+	for _, role := range roles.Items {
+		err = k8sClient.RbacV1().Roles(sourceNS).Delete(context.TODO(), role.Name, metav1.DeleteOptions{})
+		if err != nil {
+			log.Error(err, fmt.Sprintf("failed to delete roles for namespace: %s", sourceNS))
+		}
+	}
+
+	// List all the roles bindings created for ArgoCD using the label selector
+	roleBindings, err := k8sClient.RbacV1().RoleBindings(sourceNS).List(context.TODO(), metav1.ListOptions{LabelSelector: labels.Set(labelSelector.MatchLabels).String()})
+	if err != nil {
+		log.Error(err, fmt.Sprintf("failed to list role bindings for namespace: %s", sourceNS))
+		return err
+	}
+
+	// Delete all the retrieved role bindings
+	for _, roleBinding := range roleBindings.Items {
+		err = k8sClient.RbacV1().RoleBindings(sourceNS).Delete(context.TODO(), roleBinding.Name, metav1.DeleteOptions{})
+		if err != nil {
+			log.Error(err, fmt.Sprintf("failed to delete role binding for namespace: %s", sourceNS))
+		}
+	}
+
+	// Get the cluster secret used for configuring ArgoCD
+	labelSelector = metav1.LabelSelector{MatchLabels: map[string]string{common.ArgoCDSecretTypeLabel: "cluster"}}
+	secrets, err := k8sClient.CoreV1().Secrets(ownerNS).List(context.TODO(), metav1.ListOptions{LabelSelector: labels.Set(labelSelector.MatchLabels).String()})
```

Comment:
I think it's better to get the secret by name. User might have configured remote clusters (which end up having a `cluster` secret-type label as well). Now think of the situation where a remote cluster has a similar namespace to the one we are going to remove from the in-cluster secret -> oops.

Reply:
Maybe we can do the cluster address check ? if it is `https://kubernetes.default.svc` or not ? 
We can't do a name check because there can be a case where user has already created the secret with the configuration using the manual workaround we provided earlier. So we can't be sure of the name ? 
Also, we are doing a similar check while creating the secret as well, so i think cluster address check should be fine?

Reply:
Yes, checking for secret being for `https://kubernetes.default.svc` (or rather, the appropriate constant) would be sufficient.

Reply:
done!